src/cpu/dyn_cache.h

```diff
@@ -710,12 +723,50 @@ static void cache_block_closing(const uint8_t *block_start, Bitu block_size);
 #endif
 
 /* Define temporary pagesize so the MPROTECT case and the regular case share as much code as possible */
-#if defined(HAVE_MPROTECT)
+#if defined(HAVE_MPROTECT) || defined(HAVE_MMAP)
 #define PAGESIZE_TEMP PAGESIZE
 #else
 #define PAGESIZE_TEMP 4096
 #endif
 
+static constexpr size_t cache_code_start_sz = CACHE_TOTAL + CACHE_MAXSIZE + PAGESIZE_TEMP - 1 + PAGESIZE_TEMP;
+static constexpr size_t cache_blocks_sz = CACHE_BLOCKS * sizeof(CacheBlock);
+
+static inline void dyn_mem_execute()
+{
+#if defined(HAVE_PTHREAD_WRITE_PROTECT_NP)
+    if (__builtin_available(macOS 11.0, *)) {
```

Comment:
`__builtin_available` is a Clang-only language extension - I wonder if there's something more portable (so GCC can use it too?). 

I wonder how light-weight these checks are? Maybe we should run this check once on start up and save it in a const bool  (in a global? or in existing memory or cache structure?).

Reply:
If we moved this check to a one-time-on-startup function, we could use other techniques to programmatically get the macOS version -- https://stackoverflow.com/a/24004701, and this would solve let GCC continue to build running binaries on macOS.

Reply:
I had lifted this approach from [qemu](https://github.com/qemu/qemu/blob/e58c7a3bba3076890592f02d2b0e596bf191b5c2/include/qemu/osdep.h) because of the simplicity, so it may pay for me to rethink it if it fundamentally needs to be a different approach, rather than just continuing to add band-aids.

Reply:
brew install gcc, and then `CC=gcc CXX=g++ meson setup build/gcc` then `meson compile -C build/gcc`

Reply:
I'll grab gcc@9 just to line up with CI.

Reply:
I've fixed the issue by adding `HAVE_BUILTIN_AVAILABLE` to Meson, but GCC 11 (I can't use 9 due to x64-86 only) is throwing an error for `paging.cpp`:

```
ninja: Entering directory `build/gcc'
[6/170] Linking static target src/debug/libdebug.a
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive library: src/debug/libdebug.a the table of contents is empty (no object file members in the library define global symbols)
[7/170] Compiling C++ object src/cpu/libcpu.a.p/paging.cpp.o
FAILED: src/cpu/libcpu.a.p/paging.cpp.o 
g++-11 -Isrc/cpu/libcpu.a.p -Isrc/cpu -I../../src/cpu -I../../include -I. -I../.. -I/opt/homebrew/include/SDL2 -fdiagnostics-color=always -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -g -Weffc++ -Wextra-semi -D_THREAD_SAFE -MD -MQ src/cpu/libcpu.a.p/paging.cpp.o -MF src/cpu/libcpu.a.p/paging.cpp.o.d -o src/cpu/libcpu.a.p/paging.cpp.o -c ../../src/cpu/paging.cpp
/var/folders/xn/wy61l2td2wb5nc_zjmdryplw0000gn/T//cc7vmrcM.s:5601:2: error: addend too big for relocation
        adrp    x0, _paging@PAGE+37748776
```

Doesn't seem related to any of the changes I've made, and the CI gcc version doesn't seem to have an issue with this, but I just wanted to post the info.